tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,457 @@
+use crate::{
+    field::RecordFields,
+    fmt::{format, FormatEvent, FormatFields, MakeWriter},
+    layer::{Context, Layer},
+    registry::{LookupSpan, SpanRef},
+};
+use std::{any::TypeId, cell::RefCell, fmt, io, marker::PhantomData};
+use tracing_core::{
+    span::{Attributes, Id, Record},
+    Event, Subscriber,
+};
+
+/// A [`Layer`] that logs formatted representations of `tracing` events.
+///
+/// [`Layer`]: ../trait.Layer.html
+#[derive(Debug)]
+pub struct FmtLayer<
+    S,
+    N = format::DefaultFields,
+    E = format::Format<format::Full>,
+    W = fn() -> io::Stdout,
+> {
+    make_writer: W,
+    fmt_fields: N,
+    fmt_event: E,
+    _inner: PhantomData<S>,
+}
+
+/// A builder for [`FmtLayer`](struct.FmtLayer.html) that logs formatted representations of `tracing`
+/// events.
+#[derive(Debug)]
+pub struct Builder<
+    S,
+    N = format::DefaultFields,
+    E = format::Format<format::Full>,
+    W = fn() -> io::Stdout,
+> {
+    fmt_fields: N,
+    fmt_event: E,
+    make_writer: W,
+    _inner: PhantomData<S>,
+}
+
+impl<S> FmtLayer<S> {
+    /// Returns a new [`Builder`](struct.Builder.html) for configuring a `FmtLayer`.
+    pub fn builder() -> Builder<S> {
+        Builder::default()
+    }
+}
+
+// This, like the MakeWriter block, needs to be a seperate impl block because we're
+// overriding the `E` type parameter with `E2`.
+impl<S, N, E, W> Builder<S, N, E, W>
```

Comment:
This is nit-picky, but can we try to use descriptive names for some of these? `E` is particularly unfortunate, as it is usually used to refer to an error type, whereas here it is used for `FormatEvents`. `F` has a similar issue (usually a `Fn`). `Fmt` maybe? `EF` if you _really_ want short names.

Reply:
@jonhoo i'm fine with changing them; I think these letters were chosen primarily because the existing `FmtSubscriber` builder uses them, though. Probably they all should be changed to be more descriptive.